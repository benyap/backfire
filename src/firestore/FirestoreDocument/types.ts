import type { Timestamp } from "@google-cloud/firestore/build/src/timestamp";
import type { GeoPoint } from "@google-cloud/firestore/build/src/geo-point";

import type { Query } from "@google-cloud/firestore/build/src/reference/query";
import type { QueryOptions } from "@google-cloud/firestore/build/src/reference/query-options";
import type { DocumentReference } from "@google-cloud/firestore/build/src/reference/document-reference";

import * as protos from "@google-cloud/firestore/build/protos/firestore_v1_proto_api";

import api = protos.google.firestore.v1;

export interface SerializedFirestoreDocument<T = any> {
  /** The document path in Firestore. */
  path: string;
  /** The serialized document data. */
  data: T;
  /** Paths to fields that were Firestore timestamps. */
  timestamps?: string[];
  /** Paths to fields that were Firestore geopoints. */
  geopoints?: string[];
  /** Paths to fields that were Firestore document references. */
  documents?: string[];
  /** Paths to fields that were Firestore collection references and queries. */
  queries?: string[];
}

export interface DeserializedFirestoreDocument<T = any> {
  /** The document path in Firestore. */
  path: string;
  /** The deserialized document data. */
  data: T;
}

/**
 * Internal implementation: {@link Timestamp}
 */
export interface SerializedTimestamp {
  _seconds: number;
  _nanoseconds: number;
}

/**
 * Internal implementation: {@link GeoPoint}
 */
export interface SerializedGeoPoint {
  _latitude: number;
  _longitude: number;
}

/**
 * Internal implementation: {@link DocumentReference}
 */
export interface SerializedDocumentReference {
  _firestore: _Firestore;
  _path: _Path;
  // _converter: any;
}

/**
 * Internal implementation: {@link Query}
 */
export interface SerializedQuery {
  _firestore: _Firestore;
  /** Internal implementation: {@link QueryOptions} */
  _queryOptions: {
    parentPath: _Path;
    collectionId: string;
    // converter: any;
    // allDescendants: boolean;
    /** @deprecated */
    fieldFilters?: _FieldFilter[];
    filters?: _FieldFilter[];
    fieldOrders: _FieldOrder[];
    startAt?: _QueryCursor;
    endAt?: _QueryCursor;
    limit?: number;
    limitType?: _LimitType;
    offset?: number;
    projection?: api.StructuredQuery.IProjection;
    // kindless: boolean;
    // requireConsistency: boolean;
  };
  // _serializer: { allowUndefined: boolean };
  // _allowUndefined: boolean;
}

type _Firestore = { projectId: string };
type _Path = { segments: string[] };

type _FieldFilter = {
  field: _Path;
  op: api.StructuredQuery.FieldFilter.Operator;
  value: any;
};

type _FieldOrder = {
  field: _Path;
  direction: api.StructuredQuery.Direction;
};

type _QueryCursor = {
  before: boolean;
  values: api.IValue[];
};

enum _LimitType {
  First = 0,
  Last = 1,
}

export type FirestoreProtoValueDefinition = api.IValue;
export type FirestoreProtoValue =
  | string
  | number
  | boolean
  | any[]
  | { [key: string]: any }
  | Timestamp
  | Uint8Array
  | GeoPoint
  | null;
