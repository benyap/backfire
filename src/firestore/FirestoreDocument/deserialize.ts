import {
  Firestore,
  Timestamp,
  GeoPoint,
  DocumentReference,
  CollectionReference,
  Query,
} from "@google-cloud/firestore";
import { EError } from "exceptional-errors";

import {
  DeserializedFirestoreDocument,
  SerializedFirestoreDocument,
  SerializedTimestamp,
  SerializedGeoPoint,
  SerializedDocumentReference,
  SerializedQuery,
  FirestoreProtoValueDefinition,
  FirestoreProtoValue,
} from "./types";

import { getValueByPath, setValueByPath } from "~/utils";

export class DeserializationError extends EError {}

/**
 * Deserialize JSON into a Firestore document.
 *
 * @param document The serialized document.
 * @param firestore The current Firestore instance.
 * @returns The deserialized document data.
 */
export function deserializeDocument(
  document: Partial<SerializedFirestoreDocument>,
  firestore: Firestore,
): DeserializedFirestoreDocument {
  try {
    if (typeof document.path !== "string") throw new Error("missing path");
    if (typeof document.data !== "object") throw new Error("missing data");

    if (document.timestamps) {
      for (const path of document.timestamps) {
        const data = getValueByPath<SerializedTimestamp>(document.data, path);
        setValueByPath(
          document.data,
          path,
          deserializeFirestoreTimestamp(data),
        );
      }
      delete document.timestamps;
    }

    if (document.geopoints) {
      for (const path of document.geopoints) {
        const data = getValueByPath<SerializedGeoPoint>(document.data, path);
        setValueByPath(document.data, path, deserializeFirestoreGeopoint(data));
      }
      delete document.geopoints;
    }

    if (document.documents) {
      for (const path of document.documents) {
        const data =
          getValueByPath<SerializedDocumentReference>(document.data, path) ??
          {};
        setValueByPath(
          document.data,
          path,
          deserializeFirestoreDocumentReference(firestore, data),
        );
      }
      delete document.documents;
    }

    if (document.queries) {
      for (const path of document.queries) {
        const data = getValueByPath<SerializedQuery>(document.data, path) ?? {};
        setValueByPath(
          document.data,
          path,
          deserializeFirestoreQuery(firestore, data),
        );
      }
      delete document.queries;
    }

    return document as DeserializedFirestoreDocument;
  } catch (cause) {
    throw new DeserializationError("error deserializing document", {
      cause: cause as Error,
      info: document,
    });
  }
}

/**
 * Deserialize a Firestore Timestamp object.
 *
 * @param data The data to deserialize.
 * @returns The deserialized object if it was valid, `null` otherwise.
 */
function deserializeFirestoreTimestamp(
  data?: Partial<SerializedTimestamp>,
): Timestamp | null {
  const { _seconds, _nanoseconds } = data ?? {};
  if (_seconds !== undefined && _nanoseconds !== undefined)
    return new Timestamp(_seconds, _nanoseconds);
  return null;
}

/**
 * Deserialize a Firestore Geopoint object.
 *
 * @param data The data to deserialize.
 * @returns The deserialized object if it was valid, `null` otherwise.
 */
function deserializeFirestoreGeopoint(
  data?: Partial<SerializedGeoPoint>,
): GeoPoint | null {
  const { _latitude, _longitude } = data ?? {};
  if (_latitude !== undefined && _longitude !== undefined)
    return new GeoPoint(_latitude, _longitude);
  return null;
}

/**
 * Deserialize a Firestore Document Reference object.
 *
 * @param firestore The current Firestore instance.
 * @param data The data to deserialize.
 * @returns The deserialized object if it was valid. `null` otherwise.
 */
function deserializeFirestoreDocumentReference(
  firestore: Firestore,
  data?: Partial<SerializedDocumentReference>,
): DocumentReference | null {
  const { _path } = data ?? {};
  if (_path && _path.segments && _path.segments.length > 0)
    return firestore.doc(_path.segments.join("/"));
  return null;
}

/**
 * Deserialize a Firestore Query object.
 *
 * @param firestore The current Firestore instance.
 * @param data The data to deserialize.
 * @returns The deserialized object if it was value. `null` otherwise.
 */
function deserializeFirestoreQuery(
  firestore: Firestore,
  data?: Partial<SerializedQuery>,
): Query | CollectionReference | null {
  if (!data?._queryOptions) return null;

  const {
    parentPath,
    collectionId,
    // `fieldFilters` has been renamed to `filters`
    fieldFilters = [],
    filters = fieldFilters,
    fieldOrders = [],
    startAt,
    endAt,
    limit,
    limitType,
    offset,
  } = data._queryOptions;

  const path = `${parentPath.segments.join("/")}/${collectionId}`;
  let ref: Query = firestore.collection(path);

  filters.forEach(({ field, op, value }) => {
    if (op === "OPERATOR_UNSPECIFIED") return;
    const path = field.segments.join(".");
    switch (op) {
      case "ARRAY_CONTAINS":
        ref = ref.where(path, "array-contains", value);
        break;
      case "ARRAY_CONTAINS_ANY":
        ref = ref.where(path, "array-contains-any", value);
        break;
      case "EQUAL":
        ref = ref.where(path, "==", value);
        break;
      case "GREATER_THAN":
        ref = ref.where(path, ">", value);
        break;
      case "GREATER_THAN_OR_EQUAL":
        ref = ref.where(path, ">=", value);
        break;
      case "IN":
        ref = ref.where(path, "in", value);
        break;
      case "LESS_THAN":
        ref = ref.where(path, "<", value);
        break;
      case "LESS_THAN_OR_EQUAL":
        ref = ref.where(path, "<=", value);
        break;
      case "NOT_EQUAL":
        ref = ref.where(path, "!=", value);
        break;
      case "NOT_IN":
        ref = ref.where(path, "not-in", value);
        break;
    }
  });

  fieldOrders.forEach(({ field, direction }) => {
    if (direction === "DIRECTION_UNSPECIFIED") return;
    ref = ref.orderBy(
      field.segments.join("."),
      direction === "ASCENDING" ? "asc" : "desc",
    );
  });

  if (startAt) {
    const values = startAt.values.map((v) => _deserializeFirestoreValue(v));
    if (startAt.before) ref = ref.startAt(...values);
    else ref = ref.startAfter(...values);
  }

  if (endAt) {
    const values = endAt.values.map((v) => _deserializeFirestoreValue(v));
    if (endAt.before) ref = ref.endBefore(...values);
    else ref = ref.endAt(...values);
  }

  if (typeof limit === "number" && typeof limitType === "number") {
    if (limitType === 0) ref = ref.limit(limit);
    else ref = ref.limitToLast(limit);
  }

  if (typeof offset === "number") {
    ref = ref.offset(offset);
  }

  return ref;
}

/**
 * Deserialize a protobuf definition value.
 */
function _deserializeFirestoreValue(
  value: FirestoreProtoValueDefinition,
): FirestoreProtoValue {
  if ("nullValue" in value) return null;
  if ("booleanValue" in value) return value.booleanValue ?? null;
  if ("integerValue" in value) return value.integerValue ?? null;
  if ("doubleValue" in value) return value.doubleValue ?? null;
  if ("timestampValue" in value) {
    const { seconds, nanos } = value.timestampValue ?? {};
    if (typeof seconds !== "number" || typeof nanos !== "number") return null;
    return new Timestamp(seconds, nanos);
  }
  if ("stringValue" in value) return value.stringValue ?? null;
  if ("bytesValue" in value) return value.bytesValue ?? null;
  if ("referenceValue" in value) return value.referenceValue ?? null;
  if ("geoPointValue" in value) {
    const { latitude, longitude } = value.geoPointValue ?? {};
    if (typeof latitude !== "number" || typeof longitude !== "number")
      return null;
    return new GeoPoint(latitude, longitude);
  }
  if ("arrayValue" in value) {
    const { values = [] } = value.arrayValue ?? {};
    return values?.map((v) => _deserializeFirestoreValue(v)) ?? null;
  }
  if ("mapValue" in value) {
    const { fields } = value.mapValue ?? {};
    if (!fields) return null;
    return Object.keys(fields).reduce(
      (object, field) => {
        object[field] = _deserializeFirestoreValue(fields[field]!);
        return object;
      },
      {} as { [key: string]: any },
    );
  }
  return null;
}
