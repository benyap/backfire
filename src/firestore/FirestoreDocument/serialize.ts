import { SerializedFirestoreDocument } from "./types";
import {
  isFirestoreDocumentReference,
  isFirestoreGeoPoint,
  isFirestoreQuery,
  isFirestoreTimestamp,
} from "./utils";

import { deleteFieldByPath } from "~/utils";

/**
 * Serialize a Firestore document into JSON.
 *
 * @param path The path of the document to serialize.
 * @param data The document data to serialize.
 * @returns The serialized document, or `null` if `data` was not a valid document.
 */
export function serializeDocument(
  path: string,
  data: unknown,
): SerializedFirestoreDocument | null {
  const json = toJSON(data);
  const document: SerializedFirestoreDocument = { path, data: json };

  // Return `null` for any non-object values
  if (json === null || typeof json !== "object") return null;

  const { timestamps, geopoints, documents, queries } =
    findFirestoreFields(json);

  // Strip unnecessary fields from Firestore objects
  documents.forEach((path) => {
    deleteFieldByPath(json, `${path}._firestore`);
    deleteFieldByPath(json, `${path}._converter`);
  });
  queries.forEach((path) => {
    deleteFieldByPath(json, `${path}._firestore`);
    deleteFieldByPath(json, `${path}._serializer`);
    deleteFieldByPath(json, `${path}._allowUndefined`);
    deleteFieldByPath(json, `${path}._queryOptions.allDescendants`);
    deleteFieldByPath(json, `${path}._queryOptions.converter`);
    deleteFieldByPath(json, `${path}._queryOptions.filters.*.serializer`);
    deleteFieldByPath(json, `${path}._queryOptions.kindless`);
    deleteFieldByPath(json, `${path}._queryOptions.projection`);
    deleteFieldByPath(json, `${path}._queryOptions.requireConsistency`);
  });

  // Add field paths to document
  if (timestamps.length > 0) document.timestamps = timestamps;
  if (geopoints.length > 0) document.geopoints = geopoints;
  if (documents.length > 0) document.documents = documents;
  if (queries.length > 0) document.queries = queries;

  return document;
}

function toJSON(object: any) {
  let output: any;

  if (Array.isArray(object)) output = [...object];
  else if (object === null) output = null;
  else output = { ...object };

  for (const key in output) {
    const value = output[key];
    if (typeof value !== "object") continue;
    output[key] = toJSON(value);
  }

  return output;
}

/**
 * Recursively checks through the values of an object and
 * returns the paths of any Firestore values it finds.
 *
 * @param object The object to search.
 * @param path The current path in an object. Do not use for top level objects.
 * @returns The paths to any Timestamp, GeoPoint or DocumentReference fields in the object.
 */
function findFirestoreFields(
  object: { [key: string]: any },
  path: string[] = [],
) {
  let timestamps: string[] = [];
  let geopoints: string[] = [];
  let documents: string[] = [];
  let queries: string[] = [];

  Object.keys(object).forEach((field) => {
    const data = object[field];

    // Skip null/undefined values
    if (data === null || typeof data === "undefined") return;

    // Check objects
    if (typeof data === "object") {
      // Check if properties match a Firestore field
      if (isFirestoreTimestamp(data)) {
        timestamps.push([...path, field].join("."));
      } else if (isFirestoreGeoPoint(data)) {
        geopoints.push([...path, field].join("."));
      } else if (isFirestoreDocumentReference(data)) {
        documents.push([...path, field].join("."));
      } else if (isFirestoreQuery(data)) {
        queries.push([...path, field].join("."));
      }

      // Otherwise, check object fields
      else if (typeof data === "object") {
        const result = findFirestoreFields(data, [...path, field]);
        timestamps = timestamps.concat(result.timestamps);
        geopoints = geopoints.concat(result.geopoints);
        documents = documents.concat(result.documents);
        queries = queries.concat(result.queries);
      }
    }
  });

  return { timestamps, geopoints, documents, queries };
}
