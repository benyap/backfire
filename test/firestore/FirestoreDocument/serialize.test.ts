import { describe, test, expect, beforeAll } from "vitest";
import { Firestore, Timestamp, GeoPoint } from "@google-cloud/firestore";

import { serializeDocument } from "~/firestore/FirestoreDocument/serialize";

describe(serializeDocument.name, () => {
  let firestore: Firestore;

  beforeAll(() => {
    firestore = new Firestore({ projectId: "demo" });
  });

  test("serializes plain documents", () => {
    const doc = serializeDocument("email/1", {
      from: "example@email.com",
      to: ["recipient1@email.com", "recipient2@email.com"],
      content: "Dear John,\n\nHello world!",
      attachments: 2,
      tags: { work: true, important: true },
      empty: null,
    });
    expect(doc).toMatchSnapshot();
  });

  test("serializes documents with timestamps", () => {
    const doc = serializeDocument("email/1", {
      from: "example@email.com",
      date: Timestamp.fromDate(new Date("2022-01-01")),
      tags: {
        created: Timestamp.fromDate(new Date("2022-01-01")),
        dates: [Timestamp.fromDate(new Date("2022-01-01"))],
      },
    });
    expect(doc).toMatchSnapshot();
  });

  test("serializes documents with geopoints", () => {
    const doc = serializeDocument("email/1", {
      from: "example@email.com",
      location: new GeoPoint(10, 20),
      tags: {
        location: new GeoPoint(10, 20),
        list: [new GeoPoint(10, 20)],
      },
    });
    expect(doc).toMatchSnapshot();
  });

  test("serializes documents with document references", () => {
    const doc = serializeDocument("email/1", {
      from: "example@email.com",
      parent: firestore.doc("email/parent"),
      references: [firestore.doc("email/parent")],
    });
    expect(doc).toMatchSnapshot();
  });

  test("serializes documents with collection references", () => {
    const doc = serializeDocument("email/1", {
      from: "example@email.com",
      root: firestore.collection("email"),
      threads: [firestore.collection("email/1/threads")],
    });
    expect(doc).toMatchSnapshot();
  });

  test("serializes documents with queries", () => {
    const doc = serializeDocument("email/1", {
      from: "example@email.com",
      threads: firestore
        .collection("email/1/threads")
        .where("field", "array-contains", [1, 2])
        .orderBy("date", "desc"),
    });
    expect(doc).toMatchSnapshot();
  });
});
