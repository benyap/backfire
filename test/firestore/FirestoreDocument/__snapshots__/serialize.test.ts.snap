// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serializeDocument > serializes documents with collection references 1`] = `
{
  "data": {
    "from": "example@email.com",
    "root": {
      "_queryOptions": {
        "collectionId": "email",
        "endAt": undefined,
        "fieldOrders": [],
        "filters": [],
        "limit": undefined,
        "limitType": undefined,
        "offset": undefined,
        "parentPath": {
          "segments": [],
        },
        "startAt": undefined,
      },
      "_queryUtil": {
        "_firestore": {
          "_backoffSettings": {
            "backoffFactor": 1.3,
            "initialDelayMs": 100,
            "maxDelayMs": 60000,
          },
          "_clientPool": {
            "activeClients": {},
            "clientDestructor": [Function],
            "clientFactory": [Function],
            "concurrentOperationLimit": 100,
            "failedClients": {},
            "grpcEnabled": false,
            "maxIdleClients": 1,
            "terminateDeferred": {
              "promise": {},
              "reject": [Function],
              "resolve": [Function],
            },
            "terminated": false,
          },
          "_databaseId": undefined,
          "_projectId": "demo",
          "_serializer": {
            "allowUndefined": false,
            "createInteger": [Function],
            "createReference": [Function],
          },
          "_settings": {
            "libName": "gccl",
            "libVersion": "7.11.0",
            "projectId": "demo",
            "toJSON": [Function],
          },
          "_settingsFrozen": false,
          "_traceUtil": {
            "settingsAttributes": {
              "gcp.firestore.settings.database_id": "(default)",
              "gcp.firestore.settings.host": "firestore.googleapis.com:443",
              "gcp.firestore.settings.initial_retry_delay": "0.1s",
              "gcp.firestore.settings.initial_rpc_timeout": "60s",
              "gcp.firestore.settings.max_idle_channels": 1,
              "gcp.firestore.settings.max_retry_delay": "60s",
              "gcp.firestore.settings.max_rpc_timeout": "60s",
              "gcp.firestore.settings.project_id": "demo",
              "gcp.firestore.settings.retry_delay_multiplier": "1.3",
              "gcp.firestore.settings.rpc_timeout_multiplier": "1",
              "gcp.firestore.settings.total_timeout": "600s",
              "otel.scope.name": "@google-cloud/firestore",
              "otel.scope.version": "7.11.0",
            },
            "tracer": {
              "_provider": {},
              "name": "@google-cloud/firestore",
              "options": undefined,
              "version": "7.11.0",
            },
            "tracerProvider": {},
          },
          "bulkWritersCount": 0,
          "registeredListenersCount": 0,
        },
        "_queryOptions": {
          "allDescendants": false,
          "collectionId": "email",
          "converter": {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldOrders": [],
          "filters": [],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": {
            "segments": [],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
    "threads": [
      {
        "_queryOptions": {
          "collectionId": "threads",
          "endAt": undefined,
          "fieldOrders": [],
          "filters": [],
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": {
            "segments": [
              "email",
              "1",
            ],
          },
          "startAt": undefined,
        },
        "_queryUtil": {
          "_firestore": {
            "_backoffSettings": {
              "backoffFactor": 1.3,
              "initialDelayMs": 100,
              "maxDelayMs": 60000,
            },
            "_clientPool": {
              "activeClients": {},
              "clientDestructor": [Function],
              "clientFactory": [Function],
              "concurrentOperationLimit": 100,
              "failedClients": {},
              "grpcEnabled": false,
              "maxIdleClients": 1,
              "terminateDeferred": {
                "promise": {},
                "reject": [Function],
                "resolve": [Function],
              },
              "terminated": false,
            },
            "_databaseId": undefined,
            "_projectId": "demo",
            "_serializer": {
              "allowUndefined": false,
              "createInteger": [Function],
              "createReference": [Function],
            },
            "_settings": {
              "libName": "gccl",
              "libVersion": "7.11.0",
              "projectId": "demo",
              "toJSON": [Function],
            },
            "_settingsFrozen": false,
            "_traceUtil": {
              "settingsAttributes": {
                "gcp.firestore.settings.database_id": "(default)",
                "gcp.firestore.settings.host": "firestore.googleapis.com:443",
                "gcp.firestore.settings.initial_retry_delay": "0.1s",
                "gcp.firestore.settings.initial_rpc_timeout": "60s",
                "gcp.firestore.settings.max_idle_channels": 1,
                "gcp.firestore.settings.max_retry_delay": "60s",
                "gcp.firestore.settings.max_rpc_timeout": "60s",
                "gcp.firestore.settings.project_id": "demo",
                "gcp.firestore.settings.retry_delay_multiplier": "1.3",
                "gcp.firestore.settings.rpc_timeout_multiplier": "1",
                "gcp.firestore.settings.total_timeout": "600s",
                "otel.scope.name": "@google-cloud/firestore",
                "otel.scope.version": "7.11.0",
              },
              "tracer": {
                "_provider": {},
                "name": "@google-cloud/firestore",
                "options": undefined,
                "version": "7.11.0",
              },
              "tracerProvider": {},
            },
            "bulkWritersCount": 0,
            "registeredListenersCount": 0,
          },
          "_queryOptions": {
            "allDescendants": false,
            "collectionId": "threads",
            "converter": {
              "fromFirestore": [Function],
              "toFirestore": [Function],
            },
            "endAt": undefined,
            "fieldOrders": [],
            "filters": [],
            "kindless": false,
            "limit": undefined,
            "limitType": undefined,
            "offset": undefined,
            "parentPath": {
              "segments": [
                "email",
                "1",
              ],
            },
            "projection": undefined,
            "requireConsistency": true,
            "startAt": undefined,
          },
          "_serializer": {
            "allowUndefined": false,
            "createInteger": [Function],
            "createReference": [Function],
          },
        },
      },
    ],
  },
  "path": "email/1",
  "queries": [
    "root",
    "threads.0",
  ],
}
`;

exports[`serializeDocument > serializes documents with document references 1`] = `
{
  "data": {
    "from": "example@email.com",
    "parent": {
      "_path": {
        "segments": [
          "email",
          "parent",
        ],
      },
    },
    "references": [
      {
        "_path": {
          "segments": [
            "email",
            "parent",
          ],
        },
      },
    ],
  },
  "documents": [
    "parent",
    "references.0",
  ],
  "path": "email/1",
}
`;

exports[`serializeDocument > serializes documents with geopoints 1`] = `
{
  "data": {
    "from": "example@email.com",
    "location": {
      "_latitude": 10,
      "_longitude": 20,
    },
    "tags": {
      "list": [
        {
          "_latitude": 10,
          "_longitude": 20,
        },
      ],
      "location": {
        "_latitude": 10,
        "_longitude": 20,
      },
    },
  },
  "geopoints": [
    "location",
    "tags.location",
    "tags.list.0",
  ],
  "path": "email/1",
}
`;

exports[`serializeDocument > serializes documents with queries 1`] = `
{
  "data": {
    "from": "example@email.com",
    "threads": {
      "_queryOptions": {
        "collectionId": "threads",
        "endAt": undefined,
        "fieldOrders": [
          {
            "direction": "DESCENDING",
            "field": {
              "segments": [
                "date",
              ],
            },
          },
        ],
        "filters": [
          {
            "field": {
              "segments": [
                "field",
              ],
            },
            "op": "ARRAY_CONTAINS",
            "value": [
              1,
              2,
            ],
          },
        ],
        "limit": undefined,
        "limitType": undefined,
        "offset": undefined,
        "parentPath": {
          "segments": [
            "email",
            "1",
          ],
        },
        "startAt": undefined,
      },
      "_queryUtil": {
        "_firestore": {
          "_backoffSettings": {
            "backoffFactor": 1.3,
            "initialDelayMs": 100,
            "maxDelayMs": 60000,
          },
          "_clientPool": {
            "activeClients": {},
            "clientDestructor": [Function],
            "clientFactory": [Function],
            "concurrentOperationLimit": 100,
            "failedClients": {},
            "grpcEnabled": false,
            "maxIdleClients": 1,
            "terminateDeferred": {
              "promise": {},
              "reject": [Function],
              "resolve": [Function],
            },
            "terminated": false,
          },
          "_databaseId": undefined,
          "_projectId": "demo",
          "_serializer": {
            "allowUndefined": false,
            "createInteger": [Function],
            "createReference": [Function],
          },
          "_settings": {
            "libName": "gccl",
            "libVersion": "7.11.0",
            "projectId": "demo",
            "toJSON": [Function],
          },
          "_settingsFrozen": false,
          "_traceUtil": {
            "settingsAttributes": {
              "gcp.firestore.settings.database_id": "(default)",
              "gcp.firestore.settings.host": "firestore.googleapis.com:443",
              "gcp.firestore.settings.initial_retry_delay": "0.1s",
              "gcp.firestore.settings.initial_rpc_timeout": "60s",
              "gcp.firestore.settings.max_idle_channels": 1,
              "gcp.firestore.settings.max_retry_delay": "60s",
              "gcp.firestore.settings.max_rpc_timeout": "60s",
              "gcp.firestore.settings.project_id": "demo",
              "gcp.firestore.settings.retry_delay_multiplier": "1.3",
              "gcp.firestore.settings.rpc_timeout_multiplier": "1",
              "gcp.firestore.settings.total_timeout": "600s",
              "otel.scope.name": "@google-cloud/firestore",
              "otel.scope.version": "7.11.0",
            },
            "tracer": {
              "_provider": {},
              "name": "@google-cloud/firestore",
              "options": undefined,
              "version": "7.11.0",
            },
            "tracerProvider": {},
          },
          "bulkWritersCount": 0,
          "registeredListenersCount": 0,
        },
        "_queryOptions": {
          "allDescendants": false,
          "collectionId": "threads",
          "converter": {
            "fromFirestore": [Function],
            "toFirestore": [Function],
          },
          "endAt": undefined,
          "fieldOrders": [
            {
              "direction": "DESCENDING",
              "field": {
                "segments": [
                  "date",
                ],
              },
            },
          ],
          "filters": [
            {
              "field": {
                "segments": [
                  "field",
                ],
              },
              "op": "ARRAY_CONTAINS",
              "serializer": {
                "allowUndefined": false,
                "createInteger": [Function],
                "createReference": [Function],
              },
              "value": [
                1,
                2,
              ],
            },
          ],
          "kindless": false,
          "limit": undefined,
          "limitType": undefined,
          "offset": undefined,
          "parentPath": {
            "segments": [
              "email",
              "1",
            ],
          },
          "projection": undefined,
          "requireConsistency": true,
          "startAt": undefined,
        },
        "_serializer": {
          "allowUndefined": false,
          "createInteger": [Function],
          "createReference": [Function],
        },
      },
    },
  },
  "path": "email/1",
  "queries": [
    "threads",
  ],
}
`;

exports[`serializeDocument > serializes documents with timestamps 1`] = `
{
  "data": {
    "date": {
      "_nanoseconds": 0,
      "_seconds": 1640995200,
    },
    "from": "example@email.com",
    "tags": {
      "created": {
        "_nanoseconds": 0,
        "_seconds": 1640995200,
      },
      "dates": [
        {
          "_nanoseconds": 0,
          "_seconds": 1640995200,
        },
      ],
    },
  },
  "path": "email/1",
  "timestamps": [
    "date",
    "tags.created",
    "tags.dates.0",
  ],
}
`;

exports[`serializeDocument > serializes plain documents 1`] = `
{
  "data": {
    "attachments": 2,
    "content": "Dear John,

Hello world!",
    "empty": null,
    "from": "example@email.com",
    "tags": {
      "important": true,
      "work": true,
    },
    "to": [
      "recipient1@email.com",
      "recipient2@email.com",
    ],
  },
  "path": "email/1",
}
`;
